tests/lib/compiler-test-derive/src/lib.rs

```diff
@@ -0,0 +1,141 @@
+#[cfg(not(test))]
+extern crate proc_macro;
+#[cfg(not(test))]
+use ::proc_macro::TokenStream;
+#[cfg(test)]
+use ::proc_macro2::TokenStream;
+use ::quote::quote;
+#[cfg(not(test))]
+use ::syn::parse;
+#[cfg(test)]
+use ::syn::parse2 as parse;
+use ::syn::*;
+use std::path::PathBuf;
+
+mod ignores;
+
+// Reimplement parse_macro_input to use the imported `parse`
+// function. This way parse_macro_input will parse a TokenStream2 when
+// unit-testing.
+macro_rules! parse_macro_input {
+    (
+        $token_stream:ident as $T:ty
+    ) => {
+        match parse::<$T>($token_stream) {
+            Ok(data) => data,
+            Err(err) => {
+                return TokenStream::from(err.to_compile_error());
+            }
+        }
+    };
+
+    (
+        $token_stream:ident
+    ) => {
+        parse_macro_input!($token_stream as _)
+    };
+}
+
+#[proc_macro_attribute]
+pub fn compiler_test(attrs: TokenStream, input: TokenStream) -> TokenStream {
+    let path: Option<ExprPath> = parse::<ExprPath>(attrs).ok();
+    let my_fn: ItemFn = parse_macro_input!(input as ItemFn);
+    let fn_name = my_fn.sig.ident.clone();
+
+    // Let's build the ignores to append an `#[ignore]` macro to the
+    // autogenerated tests in case the test appears in the `ignores.txt` path;
+
+    let mut ignores_txt_path = PathBuf::new();
+    ignores_txt_path.push(env!("CARGO_MANIFEST_DIR"));
+    ignores_txt_path.push("../../ignores.txt");
+
+    let ignores = crate::ignores::Ignores::build_from_path(ignores_txt_path);
+
+    let should_ignore = |test_name: &str, compiler_name: &str, engine_name: &str| {
+        let compiler_name = compiler_name.to_lowercase();
+        let engine_name = engine_name.to_lowercase();
+        // We construct the path manually because we can't get the
+        // source_file location from the `Span` (it's only available in nightly)
+        let full_path = format!(
+            "{}::{}::{}::{}",
+            quote! { #path },
```

Comment:
This is surprising to me, I thought `quote!` was for making ASTs of Rust tokens; does this do the right thing? If I ran across this code in the wild, I'd be very appreciative of a comment explaining the idea here because it looks like it shouldn't work, but I'm guessing it does

Reply:
It took me a bit to get this working to be honest. It was not trivial to pass from a TokenStream to a string actually.
But yeah, I fully tested it and it's working :)